core/src/main/java/io/temporal/samples/hello/HelloAccumulator.java

```diff
@@ -0,0 +1,625 @@
+    }
+  }
+
+  // Main workflow method
+  public static class AccumulatorWorkflowImpl implements AccumulatorWorkflow {
+
+    private final GreetingActivities activities =
+        Workflow.newActivityStub(
+            GreetingActivities.class,
+            ActivityOptions.newBuilder().setStartToCloseTimeout(Duration.ofSeconds(2)).build());
+
+    private static final Logger logger = LoggerFactory.getLogger(AccumulatorWorkflowImpl.class);
+    String bucketKey;
+    ArrayDeque<Greeting> greetings;
+    HashSet<String> allGreetingsSet;
+    boolean stopWaiting = false;
+    ArrayDeque<Greeting> unprocessedGreetings;
+
+    @Override
+    public String accumulateGreetings(
+        String bucketKeyInput,
+        ArrayDeque<Greeting> greetingsInput,
+        Set<String> allGreetingsSetInput) {
+      bucketKey = bucketKeyInput;
+      if (unprocessedGreetings == null) unprocessedGreetings = new ArrayDeque<Greeting>();
+      greetings = new ArrayDeque<Greeting>();
+      allGreetingsSet = new HashSet<String>();
+      greetings.addAll(greetingsInput);
+      allGreetingsSet.addAll(allGreetingsSetInput);
+
+      // If you want to wait for a fixed amount of time instead of a time after a message
+      // as this does now, you might want to check out
+      // ../../updatabletimer
+
+      // Main Workflow Loop:
+      //    - wait for signals to come in
+      //    - every time a signal comes in, wait again for MAX_AWAIT_TIME
+      //    - if time runs out, and there are no messages, process them all and exit
+      //    - if exit signal is received, process any remaining signals and exit
+      do {
+
+        Workflow.await(MAX_AWAIT_TIME, () -> !unprocessedGreetings.isEmpty() || stopWaiting);
+
+        if (!unprocessedGreetings.isEmpty()) {
+          processGreeting(unprocessedGreetings.removeFirst());
+        } else {
+          String greetEveryone = processGreetings(greetings);
+
+          if (unprocessedGreetings.isEmpty()) {
+            logger.info("Greeting queue is still empty");
+            return greetEveryone;
+          } else if (stopWaiting && unprocessedGreetings.isEmpty()) {
+            return greetEveryone;
+          } else {
+            // you can get here if you send a signal after an exit, causing rollback just after the
+            // last processed activity
+            logger.info("Greeting queue not empty, looping");
+          }
+        }
+      } while (!unprocessedGreetings.isEmpty() || !Workflow.getInfo().isContinueAsNewSuggested());
+
+      logger.info("starting continue as new processing");
+
+      // Create a workflow stub that will be used to continue this workflow as a new
+      AccumulatorWorkflow continueAsNew = Workflow.newContinueAsNewStub(AccumulatorWorkflow.class);
+
+      // Request that the new run will be invoked by the Temporal system:
+      continueAsNew.accumulateGreetings(bucketKey, greetings, allGreetingsSet);
```

Comment:
you could get buffered signal(s) at this point, guess need to decide to handle them or send them to next exec

Reply:
Thanks for the comment!

Is this in the case of a high volume of signals, so that one would come in between the do loop above ending and the continue as new happening?

Do we have sample code that shows how to work with buffered signals?

Reply:
We could likely make an improvement to this by using our upcoming are_handlers_finished API, once it's ready.

Reply:
@joshmsmith as I understand it, you're doing it correctly.  The server will fail the workflow task if any signals come in after this, and the task will be retried.

Aside: We're adding a predicate called something like `allMessagesHandled` that you'll be able to check, but it's more for people that aren't already enqueueing requests.